src/r1cs.rs

```diff
@@ -31,6 +31,7 @@ pub struct R1CSShape<G: Group> {
   pub(crate) num_cons: usize,
   pub(crate) num_vars: usize,
   pub(crate) num_io: usize,
+  pub(crate) constraints_path: Option<Vec<String>>,
```

Comment:
I opt this part of memory usage, by 
1. `is_sat/is_sat_relaxed` will return just `index: usize` of constraint error
2. then the `usize -> String` will be happened on verify() function.
3. There is no `ShapeCS` reference holding and carry over, instead the `constraint_path: Vec<String>` is put in `PublicParams` and setup only once
4. remove  `derived(clone)` from PublicParams, to avoid abuse

Which  I think is the better handling for now, with minimal interface change to achieve the happy debugging

Reply:
Beside I also fine tune running instances vector copy by `to_vec()` function. I think now memory opt a lot